src/Lean/Meta/Tactic/AC/Sharing.lean

```diff
@@ -0,0 +1,303 @@
+
+abbrev VarStateM  := StateT VarState MetaM
+
+def VarStateM.run' (x : VarStateM α) (s : VarState := {}) : MetaM α :=
+  StateT.run' x s
+
+/-! ### Implementation -/
+
+/-- Return a list with all variable indices that have a mapping.
+
+Note that this is always a complete sequence `0, 1, ..., (n-1)`, without skipping
+numbers. -/
+def getAllVarIndices : VarStateM (List VarIndex) := do
+  pure <| List.range (← get).varIndices.size
+
+/-- Return the unique variable index for an expression.
+
+Modifies the monadic state to add a new mapping and increment the index,
+if needed. -/
+def VarStateM.exprToVar (e : Expr) : VarStateM VarIndex := fun state =>
+  -- TODO: we should consider normalizing `e` here using `AC.rewriteUnnormalized`, so that distinct
+  --   atomic expressions which are equal up-to associativity and commutativity of another operator
+  --   get mapped to the same variable id
+  return match state.varIndices[e]? with
+  | some idx => (idx, state)
+  | none =>
+    let { varIndices, varExprs } := state
+    let nextIndex := varIndices.size
+    let varIndices := varIndices.insert e nextIndex
+    let varExprs := varExprs.push e
+    (nextIndex, { varIndices, varExprs })
+
+/-- Return the expression that is represented by a specific variable index. -/
+def VarStateM.varToExpr (idx : VarIndex) : VarStateM Expr := do
+  let { varExprs, .. } ← get
+  if h : idx < varExprs.size then
+    pure varExprs[idx]
+  else
+    throwError "internal error (this is a bug!): index {idx} out of range, \
+      the current state only has {varExprs.size} variables:\n\n{varExprs}"
+
+/-- Given a binary, associative and commutative operation `op`,
+decompose expression `e` into its variable coefficients.
+
+For example `a ⬝ b ⬝ (a ⋅ c)` will give the coefficients:
+```
+a => 2
+b => 1
+c => 1
+```
+
+Any compound expression which is not an application of the given `op`
+-/
+def VarStateM.computeCoefficients (op : Expr) (e : Expr) : VarStateM CoefficientsMap :=
+  go {} e
+where
+  incrVar (coe : CoefficientsMap) (e : Expr) : VarStateM CoefficientsMap := do
+    let idx ← exprToVar e
+    return coe.alter idx (fun c => some <| (c.getD 0) + 1)
+  go (coe : CoefficientsMap) : Expr → VarStateM CoefficientsMap
+  | e@(AC.bin op' x y) => do
+      if ← isDefEq op op' then
+        let coe ← go coe x
+        let coe ← go coe y
+        return coe
+      else
+        incrVar coe e
+  | e => incrVar coe e
+
+structure SharedCoefficients where
+  common : CoefficientsMap := {}
+  x : CoefficientsMap
+  y : CoefficientsMap
+
+/-- Given two sets of coefficients `x` and `y` (computed with the same variable
+mapping), extract the shared coefficients, such that `x` (resp. `y`) is the sum of
+coefficients in `common` and `x` (resp `y`) of the result. -/
+def SharedCoefficients.compute (x y : CoefficientsMap) : VarStateM SharedCoefficients := do
+  let mut res : SharedCoefficients := { x, y }
+
+  for idx in ← getAllVarIndices do
```

Comment:
```suggestion
  for idx in [0:(← get).varIndices.size]do
```

Way more efficient than materializing an `O(n)` size data structure using `O(n)` time to iterate from `0` to `n`!

Reply:
The current version should be fine, since I refactored `getAllVarIndices` to return the suggested range